src/main/scala/subsystem/HasTiles.scala

```diff
@@ -234,20 +258,49 @@ trait CanAttachTile {
     context.tileWFIXbarNode :=* tile.wfiNode
     context.tileCeaseXbarNode :=* tile.ceaseNode
   }
+
+  def connectInputConstants(tile: TileType, context: TileContextType): Unit = {
+    implicit val p = context.p
+    tile.hartIdNode := context.tileHartIdNode
+    tile.resetVectorNode := context.tileResetVectorNode
+  }
 }
 
-/** Provides some Chisel connectivity to certain tile IOs */
-trait HasTilesModuleImp extends LazyModuleImp with HasPeripheryDebugModuleImp {
-  val outer: HasTiles with HasTileInterruptSources
+/** InstantiatesTiles adds a Config-urable sequence of tiles of any type
+  *   to the subsystem class into which it is mixed.
+  */
+trait InstantiatesTiles { this: BaseSubsystem =>
+  // Actually instantiate all tiles, in order based on statically-assigned hartids
+  val tileAttachParams: Seq[CanAttachTile] = p(TilesLocated(location)).sortBy(_.tileParams.hartId)
+  val tiles: Seq[BaseTile] = tileAttachParams.map(_.instantiate(p))
+
+  // Helper functions for accessing certain parameters the are popular to refer to in subsystem code
+  val tileParams: Seq[TileParams] = tileAttachParams.map(_.tileParams)
+  val tileCrossingTypes = tileAttachParams.map(_.crossingParams.crossingType)
+  def nTiles: Int = tileAttachParams.size
+  def hartIdList: Seq[Int] = tileParams.map(_.hartId)
+  def localIntCounts: Seq[Int] = tileParams.map(_.core.nLocalInterrupts)
+
+  require(hartIdList.distinct.size == tiles.size, s"Every tile must be statically assigned a unique id, but got:\n${hartIdList}")
+}
+
+/** HasTiles instantiates and also connects a Config-urable sequence of tiles of any type to subsystem interconnect resources. */
+trait HasTiles extends InstantiatesTiles with HasCoreMonitorBundles with DefaultTileContextType
+{ this: BaseSubsystem => // TODO: ideally this bound would be softened to Attachable
+  implicit val p: Parameters
 
-  def resetVectorBits: Int = {
-    // Consider using the minimum over all widths, rather than enforcing homogeneity
-    val vectors = outer.tiles.map(_.module.constants.reset_vector)
-    require(vectors.tail.forall(_.getWidth == vectors.head.getWidth))
-    vectors.head.getWidth
+  // connect all the tiles to interconnect attachment points made available in this subsystem context
+  tileAttachParams.zip(tiles).foreach { case (params, t) =>
+    params.connect(t.asInstanceOf[params.TileType], this.asInstanceOf[params.TileContextType])
   }
+}
+
+/** Provides some Chisel connectivity to certain tile IOs */
+trait HasTilesModuleImp extends LazyModuleImp with HasPeripheryDebugModuleImp {
+  val outer: HasTiles with HasTileInterruptSources with HasTileInputConstants
 
-  val tile_inputs = outer.tiles.map(_.module.constants)
+  val reset_vector = outer.tileResetVectorIONodes.zipWithIndex.map { case (n, i) => n.makeIO(s"reset_vector_$i") }
+  val tile_hartids = outer.tileHartIdIONodes.zipWithIndex.map { case (n, i) => n.makeIO(s"tile_hartids_$i") }
```

Comment:
Right so if you're going to suggest names of IOs, Chisel requires you to provide unique names (it will not do `_#` suffixing for you). Now since these values are top-level vals in the `LazyModuleImp`, you could let Chisel reflectively name them rather them rather than using `ValName`. The reason why you're having to provide the name yourself is because `ValName` just picks the name of the val, it assumes Chisel will take care of the suffixiing.